### PR TITLE
1376 dev building matching agent

### PR DESCRIPTION
**Pull request by markushofmeister**
I have tested this agent for the latest instantiated EPC data and it works as expected. To avoid having this branch being stale for too long, I suggest to merge it to main

### CODE REVIEW

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  <groupId>uk.ac.cam.cares.jps</groupId>
+  <artifactId>BuildingMatchingAgent</artifactId>
+  <version>1.0.0</version>
+  <packaging>war</packaging>
+
+  <!-- Parent POM -->
+  <parent>
+    <groupId>uk.ac.cam.cares.jps</groupId>
+    <artifactId>jps-parent-pom</artifactId>
+    <version>1.0.0</version>
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml:16]: Please update this to the latest version.

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  </parent>
+
+  <name>BuildingMatchingAgent Maven Webapp</name>
+  <!-- FIXME change it to the project's website -->
+  <url>http://www.example.com</url>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>1.7</maven.compiler.source>
+    <maven.compiler.target>1.7</maven.compiler.target>
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml:26]: This is woefully out of date. Please use something more recent.

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <version>3.1.0</version>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>uk.ac.cam.cares.jps</groupId>
+      <artifactId>jps-base-lib</artifactId>
+      <version>1.11.0</version>
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml:39]: This is woefully out of date. Please update to the most recent version.

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+  <build>
+    <sourceDirectory>src/main/java</sourceDirectory>
+    <testSourceDirectory>src/test/java</testSourceDirectory>
+    <pluginManagement>
+      <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.7.0</version>
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml:63]: Please update to a more recent version.

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      </plugin>
+    </plugins>
+    </pluginManagement>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <source>8</source>
+          <target>8</target>
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml:102]: Please use something more recent.

File: Agents/BuildingMatchingAgent/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# First stage: build war file
+#==================================================================================================
+FROM maven:3.6-openjdk-8-slim as builder
+
+# Copy all files into root's home, including the source, pom file, ./m2 directory, credentials and config files
+ADD . /root
+
+# Populate settings templates with credentials, repo name
+WORKDIR /root/.m2
+# (Note that | rather than / is used as the sed delimiter, since encrypted passwords can contain the former, but not the latter
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/Dockerfile:10]: "latter, but not the former)"

File: Agents/BuildingMatchingAgent/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    ports:
+      - "58085:8080"
+
+    volumes:
+      - external_maven_repo:/root/.m2
+
+
+volumes:
+  external_maven_repo:
+    name: "external_maven_repo"
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/docker-compose.yml:17]: Is this volume required? I'm guessing it's just a convenience...

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/src/main/java/uk/ac/cam/cares/jps/BuildingMatchingAgent.java [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    private static SelectBuilder ocgmlQueryBuilder() throws ParseException {
+//        WhereBuilder where = new WhereBuilder().addPrefix(KEY_OCGML, ocgmlUri).addWhere(QM+BLDG, KEY_OCGML+":objectClassId",  "26");
+//
+//        return new SelectBuilder().setDistinct(true)
+//                .addPrefix(KEY_OCGML, ocgmlUri).addPrefix("osid", osidUri)
+//                .addVar(QM+BLDG).addVar(QM+uprn)
+//                .addGraph(NodeFactory.createURI(bldgGraph), where)
+//                .addBind("IRI(REPLACE(str("+QM+BLDG+"), \"building\", \"cityobject\"))", QM+CITYOBJ)
+//                .addWhere(QM+ATTR, "osid:intersectsFeature" , QM+CITYOBJ)
+//                .addWhere(QM+ATTR, "osid:hasValue", QM+uprn).addOrderBy(QM+BLDG);
```
Comment (sm453) [inline Agents/BuildingMatchingAgent/BuildingMatchingAgent/src/main/java/uk/ac/cam/cares/jps/BuildingMatchingAgent.java:220]: I'm guessing this commented-out bit is just for testing/debugging? If so, please remove.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml (hunk 1 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml]: Addressed by commit [78556da](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/78556da93e569f13b79e10f6f7d0c027455378a4)

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml (hunk 2 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml]: Addressed by commit [78556da](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/78556da93e569f13b79e10f6f7d0c027455378a4)

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml (hunk 3 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml]: Addressed by commit [78556da](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/78556da93e569f13b79e10f6f7d0c027455378a4)

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml (hunk 4 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml]: Addressed by commit [78556da](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/78556da93e569f13b79e10f6f7d0c027455378a4)

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml (hunk 5 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/pom.xml]: Addressed by commit [78556da](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/78556da93e569f13b79e10f6f7d0c027455378a4)

File: Agents/BuildingMatchingAgent/Dockerfile (hunk 6 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/Dockerfile]: Addressed by commit [ee3cc25](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/ee3cc255c196a2ed5e1ca514823e3f49596895d9)

File: Agents/BuildingMatchingAgent/docker-compose.yml (hunk 7 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/docker-compose.yml]: Addressed by commit [ee3cc25](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/ee3cc255c196a2ed5e1ca514823e3f49596895d9)

File: Agents/BuildingMatchingAgent/BuildingMatchingAgent/src/main/java/uk/ac/cam/cares/jps/BuildingMatchingAgent.java (hunk 8 above)
Comment (markushofmeister) [reply Agents/BuildingMatchingAgent/BuildingMatchingAgent/src/main/java/uk/ac/cam/cares/jps/BuildingMatchingAgent.java]: Addressed by commit [00b2e50](https://github.com/cambridge-cares/TheWorldAvatar/pull/616/commits/00b2e50c4103dbf69dfc04feb670fbe5c9fd1617)

**Review by sm453 — APPROVED** [review]

**Pull request merged**